### PR TITLE
Modified the UUID class to use a primitive and added parsing options

### PR DESCRIPTION
**Pull request by Mathilde411**
The UUID class has been modified to use the primitive from pharo-project/pharo-vm#610.

Parsing methods were also added mainly for version 1 RFC 4122 UUIDs.

See: https://datatracker.ietf.org/doc/html/rfc4122

### CODE REVIEW

File: src/Network-UUID/UUID.class.st [hunk 1]
```diff
@@ -142,36 +135,119 @@ UUID >> asString36 [
...
+	
+	"Should be deprecated, but is still used by iceberg."
+	"self
+		deprecated:
+		'The UUIDs are no longer a subclass of ByteArray, and should not be directly modified.'."
+	^ uuidData at: aNumber
+]
+
+{ #category : #accessing }
+UUID >> at: aNumber put: aByte [
```
Comment (Mathilde411) [inline src/Network-UUID/UUID.class.st:179]: This Method should be deprecated, but is currently used by Iceberg. Will change it after Iceberg is updated.

**Comment by jecisc** [comment]
The bootstrap is failing with:

```
MetacelloNotification: Loading -> Iceberg-CompatibleUserName.1686126007 --- https://github.com/pharo-vcs/Iceberg.git[dev-2.0] --- cache
Error: You are about to load new versions of the following packages
that have unsaved changes in the image:
  Iceberg

If you continue, you will lose these changes:
MCVersionLoader>>warnAboutLosingChangesTo:ifCancel:ifMerge:
MCVersionLoader>>checkForModificationsIfCancel:ifMerge:
MCVersionLoader>>basicLoadWithNameLike:
```

This error is a new one to me on the CI

File: src/Network-Tests/UUIDGeneratorTest.class.st [hunk 2]
```diff
@@ -18,67 +18,75 @@ UUIDGeneratorTest >> setUp [
...
+	self assert: UUIDGenerator default equals: UUIDGenerator default.
 ]
 
 { #category : #testing }
 UUIDGeneratorTest >> testSpeed [
+	|array|
+	array := ByteArray new: 16.
 	self
-		should: [ 100000 timesRepeat: [ generator next ] ]
+		should: [ 100000 timesRepeat: [ UUIDGenerator default placeFields: array ] ]
```
Comment (guillep) [inline src/Network-Tests/UUIDGeneratorTest.class.st]: I don't like this test.

First, it's breaking the public API. Why testing `placeFields:` and not directly `UUID new`.
Then, testing speed like this is just a smoke test. Why 100000 and not more (or less?) :D.
I'd remove this test altogether.

File: src/Fuel-Tests-Core/FLBasicSerializationTest.class.st [hunk 3]
```diff
@@ -796,7 +796,7 @@ FLBasicSerializationTest >> testTrue [
 FLBasicSerializationTest >> testUUID [
 
 	self assertSerializationEqualityOf: (UUID fromString: 'a3b64357-377a-5b41-b575-1c653084a121').
-	self assertSerializationEqualityOf: UUID nilUUID.
+	self assertSerializationEqualityOf: NilUUID new.
```
Comment (guillep) [inline src/Fuel-Tests-Core/FLBasicSerializationTest.class.st]: What is the rationale behind NilUUID ?
Why a different class?

If having a nil uuid is really important, why couldn't we have `UUID nilUUID`? That way the implementation is kinda hidden (one class, two classes, the user should not care, it's easier to refactor)

File: src/Network-Tests/UUIDGeneratorTest.class.st [hunk 4]
```diff
@@ -18,67 +18,75 @@ UUIDGeneratorTest >> setUp [
...
+
 	| generator1 generator2 uuid1 uuid2 time1 time2 |
 	generator1 := UUIDGenerator new.
 	generator2 := UUIDGenerator new.
-	uuid1 := generator1 next.
-	uuid2 := generator2 next.
+	uuid1 := NilUUID new.
+	uuid2 := NilUUID new.
+	generator1 placeFields: uuid1 uuidData.
+	generator2 placeFields: uuid2 uuidData.
```
Comment (guillep) [inline src/Network-Tests/UUIDGeneratorTest.class.st]: This test breaks the public interface. We should kill it probably or rewrite it :)

Let's discuss

File: src/Network-UUID/ByteArray.extension.st [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Extension { #name : #ByteArray }
+
+{ #category : #'*Network-UUID' }
+ByteArray >> generateUUIDInPlace [
+	<primitive: 'primitiveMakeUUID' module: 'UUIDPlugin'>
+	self size = 16 ifFalse: [ self error: 'The byte array size has to be 16 bytes.' ].
+	UUIDGenerator default placeFields: self
```
Comment (guillep) [inline src/Network-UUID/ByteArray.extension.st:7]: Nice!!

File: src/Network-UUID/NilUUID.class.st [hunk 6]
```diff
@@ -0,0 +1,13 @@
+"
+UUID with all fields set to 0.
```
Comment (guillep) [inline src/Network-UUID/NilUUID.class.st]: My point: We don't need a separate class for this :) Just a different instance is enough!

File: src/Network-UUID/UUID.class.st [hunk 7]
```diff
@@ -1,6 +1,5 @@
 "
 I am UUID.
-I am a ByteArray.
```
Comment (guillep) [inline src/Network-UUID/UUID.class.st:3]: Cuack :D

File: src/Network-UUID/UUID.class.st [hunk 8]
```diff
@@ -187,13 +273,69 @@ UUID >> printString [
...
+	"Returns the variant of the UUID. See https://tools.ietf.org/html/rfc4122"
+
+	^ (uuidData at: 9) >> 5
+]
+
+{ #category : #comparing }
+UUID >> version [
+	"Returns the version of the UUID. See https://tools.ietf.org/html/rfc4122"
+
+	^ (uuidData at: 7) >> 4
```
Comment (guillep) [inline src/Network-UUID/UUID.class.st:356]: It's nice to have a real UUID implementation!

**Review by guillep — CHANGES_REQUESTED** [review]
I left some comments.

I think we should make Iceberg not need to look at the internals, and we should discuss the API :)

File: src/Network-UUID/UUID.class.st (hunk 7 above)
Comment (Mathilde411) [reply src/Network-UUID/UUID.class.st:3]: 🦆

File: src/Network-UUID/NilUUID.class.st (hunk 6 above)
Comment (Mathilde411) [reply src/Network-UUID/NilUUID.class.st]: NilUUID has been removed :)

File: src/Network-Tests/UUIDGeneratorTest.class.st (hunk 4 above)
Comment (Mathilde411) [reply src/Network-Tests/UUIDGeneratorTest.class.st]: NilUUID has been removed :)

File: src/Network-Tests/UUIDGeneratorTest.class.st [hunk 9]
```diff
@@ -18,67 +18,70 @@ UUIDGeneratorTest >> setUp [
...
 	"different generators, different node id"
-	self deny: uuid1 reverse readStream uint32 equals: uuid2 reverse readStream uint32.
+	self
+		deny: uuid1 uuidData reverse readStream uint32
+		equals: uuid2 uuidData reverse readStream uint32.
 	"same generators, similar clock, within 100 milliseconds (stored time is in microseconds)"
 	time1 := self timeFromUUID: uuid1.
 	time2 := self timeFromUUID: uuid2.
-	self assert: (time2 - time1) abs < (100 * 1000)
+	"self assert: (time2 - time1) abs < (100 * 1000)"
```
Comment (guillep) [inline src/Network-Tests/UUIDGeneratorTest.class.st:65]: Should we remove the commented code?

**Review by guillep — APPROVED** [review]
I think this is ok, modulo one line of code that is commented.

Question: are you planning to change Iceberg to not access the UUID directly?

**Comment by Mathilde411** [comment]
> I think this is ok, modulo one line of code that is commented.
> 
> Question: are you planning to change Iceberg to not access the UUID directly?

I was waiting for the changes to be merged since it is not in the main repo, but yes.

File: src/Network-Tests/UUIDGeneratorTest.class.st (hunk 9 above)
Comment (guillep) [inline src/Network-Tests/UUIDGeneratorTest.class.st:65]:
```suggestion
```

File: src/Network-UUID/UUID.class.st [hunk 10]
```diff
@@ -187,13 +288,69 @@ UUID >> printString [
...
+
+{ #category : #accessing }
+UUID >> timestamp [
+	"Returns the timestamp of the version 1 UUID as a local DateAndTime. See https://tools.ietf.org/html/rfc4122"
+
+	| timestamp |
+	self isTimeBased ifFalse: [
+		self error: 'Only time-based UUIDs have a timestamp.' ].
+
+	timestamp := (uuidData at: 7) & 16r0f.
```
Comment (guillep) [inline src/Network-UUID/UUID.class.st:320]:
```suggestion
	timestamp := (uuidData at: 7) & 16r0f.
	"The timestamp bytes in the UUID bytearray are scrambled. Iterate them in the good order"
```

File: src/Network-Tests/UUIDGeneratorTest.class.st [hunk 11]
```diff
@@ -18,67 +18,70 @@ UUIDGeneratorTest >> setUp [
...
+	"UUID are not necessarly of version 1->3, time can be random"
```
Comment (guillep) [inline src/Network-Tests/UUIDGeneratorTest.class.st]:
```suggestion
```

**Comment by guillep** [comment]
CI is green! changes are good! Thanks @BastouP411

**Pull request merged**